Add std::thread::add_spawn_hook.

m-ou-se opened the pull request:
Implementation of https://github.com/rust-lang/rfcs/pull/3642

m-ou-se committed:
Add std::thread::add_spawn_hook.

m-ou-se committed:
Use add_spawn_hook for libtest's output capturing.

m-ou-se commented on library/std/src/thread/spawnhook.rs:
```diff
@@ -0,0 +1,92 @@
+use crate::sync::RwLock;
+use crate::thread::Thread;
+
+static SPAWN_HOOKS: RwLock<
+    Vec<&'static (dyn Fn(&Thread) -> io::Result<Box<dyn FnOnce() + Send>> + Sync)>,
+> = RwLock::new(Vec::new());
```
An alternative is to use a lock free linked list:

```rust
static SPAWN_HOOK: AtomicPtr<SpawnHook> = AtomicPtr::new(ptr::null_mut());

struct SpawnHook {
    next: Option<&'static SpawnHook>,
    hook: &'static (dyn Fn(&Thread) -> io::Result<Box<dyn FnOnce() + Send>> + Sync),
}

pub fn add_spawn_hook<F, G>(hook: F)
where
    F: 'static + Sync + Fn(&Thread) -> io::Result<G>,
    G: 'static + Send + FnOnce(),
{
    let mut next = SPAWN_HOOK.load(Ordering::Relaxed);
    let mut hook = Box::leak(Box::new(SpawnHook {
        next: unsafe { next.as_ref() },
        hook: Box::leak(Box::new(move |thread: &Thread| {
            let f: Box<dyn FnOnce() + Send> = Box::new(hook(thread)?);
            Ok(f)
        })),
    }));
    while let Err(n) =
        SPAWN_HOOK.compare_exchange_weak(next, hook, Ordering::Release, Ordering::Relaxed)
    {
        next = n;
        hook.next = unsafe { next.as_ref() };
    }
}
```

But that uses two Boxes per hook (one for the `dyn Fn` and one for the linked list node (`SpawnHook`). We could merge them into a single `Box` by storing the `Fn` by value in a `SpawnHook<F>`, but that results in a lot more complexity and much more subtle unsafe code, which is probably not worth it.

bors commented:
:umbrella: The latest upstream changes (presumably #127998) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

m-ou-se committed:
Update thread spawn hooks.

1. Make the effect thread local.
2. Don't return a io::Result from hooks.

m-ou-se committed:
Add thread Builder::no_hooks().

m-ou-se commented:
This is now updated to match the [updated RFC](https://github.com/rust-lang/rfcs/pull/3642#issuecomment-2434748643):

> 1. The effect is now thread-local instead of global. Adding a hook only affects future child threads of the current thread.
> 2. Removed the use of io::Result.
> 3. Added an opt-out method to the thread builder.

m-ou-se committed:
Use Send + Sync for spawn hooks.

m-ou-se committed:
Add tracking issue.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11801360432/job/32874535863) [(plain)](https://github.com/rust-lang/rust/commit/80162f727732ed1bcc55af7e697698f90b48aeca/checks/32874535863/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.926 Building wheels for collected packages: reuse
#13 2.927   Building wheel for reuse (pyproject.toml): started
#13 3.179   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.180   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#13 3.180   Stored in directory: /tmp/pip-ephem-wheel-cache-t4wcbnw8/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.183 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.589 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.589 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 4.140 Collecting virtualenv
#13 4.140 Collecting virtualenv
#13 4.191   Downloading virtualenv-20.27.1-py3-none-any.whl (3.1 MB)
#13 4.320      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 24.7 MB/s eta 0:00:00
#13 4.378 Collecting platformdirs<5,>=3.9.1
#13 4.385   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#13 4.425 Collecting filelock<4,>=3.12.2
#13 4.455 Collecting distlib<1,>=0.3.7
#13 4.462   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#13 4.476      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 41.8 MB/s eta 0:00:00
#13 4.476      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 41.8 MB/s eta 0:00:00
#13 4.559 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.738 Successfully installed distlib-0.3.9 filelock-3.16.1 platformdirs-4.3.6 virtualenv-20.27.1
#13 DONE 4.8s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      307136 kB
DirectMap2M:     9129984 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://static.rust-lang.org/dist/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/2024-10-16/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt
---
fmt check
fmt: checked 5661 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/std/src/thread/spawnhook.rs:86: feature gate thread_spawn_hook has inconsistent `issue`: "none" mismatches the previous `issue` of "132951"
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (24.3.1)
All checks passed!
checking C++ file formatting
some tidy checks failed
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Tue Nov 12 16:32:48 UTC 2024
  network time: Tue, 12 Nov 2024 16:32:48 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

bjorn3 commented on library/std/src/thread/mod.rs:
```diff
@@ -335,6 +342,16 @@ impl Builder {
+    pub fn no_hooks(mut self) -> Builder {
+        self.no_hooks = true;
+        self
+    }
```
Should this be allowed when doing scoped thread spawning? Lazy spawning in a thread pool wouldn't use scoped thread spawning and denying `no_hooks` for scoped thread spawning could potentially allow `!Send` types to depend on thread local storage existing on all threads that can access it. For example if you have a C library which needs to be initialized on every thread, you may be able to pass a token around indicating that it is initialized and install a spawn hook to initialize it on every new thread. Then a `Sync + !Send` token would allow accessing the C api on child threads without having to check for initialization. This example may be a bit contrived though.

bjorn3 commented on library/std/src/thread/spawnhook.rs:
```diff
@@ -0,0 +1,145 @@
+
+// A thread local linked list of spawn hooks.
+crate::thread_local! {
+    static SPAWN_HOOKS: Cell<SpawnHooks> = const { Cell::new(SpawnHooks { first: None }) };
```
Why does this use a linked list rather than a Vec?

m-ou-se replied:
So the Vec doesn't have to be cloned when a new thread is spawned. It just creates an extra Arc to the (current) head of the list, making it a linked 'tree' with multiple heads.

m-ou-se committed:
Fix tracking issue.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11914987774/job/33204219299) [(plain)](https://github.com/rust-lang/rust/commit/9c666b192a51bf7fbd8cef15793cd3b2f378a0ca/checks/33204219299/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 26.9s done
#16 DONE 32.8s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.91s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
```

</details>

WaffleLapkin commented on library/std/src/thread/spawnhook.rs:
```diff
@@ -0,0 +1,145 @@
+///
+/// Hooks can only be added, not removed.
+///
+/// The hooks will run in order, starting with the most recently added.
```
maybe "in reverse order"?

WaffleLapkin reviewed (commented):
a few nitpicks

m-ou-se commented on library/std/src/thread/spawnhook.rs:
```diff
@@ -0,0 +1,145 @@
+    F: 'static + Send + Sync + Fn(&Thread) -> G,
+    G: 'static + Send + FnOnce(),
+{
+    SPAWN_HOOKS.with(|h| {
+        let mut hooks = h.take();
+        hooks.first = Some(Arc::new(SpawnHook {
+            hook: Box::new(move |thread| Box::new(hook(thread))),
+            next: hooks.first.take(),
+        }));
+        h.set(hooks);
+    });
```
> but it doesn't compile because SpawnHooks impelements Drop..

Yeah. We really need a feature to destruct Drop types into their fields.

m-ou-se committed:
Address review comments.

Co-authored-by: waffle <waffle.lapkin@gmail.com>

m-ou-se committed:
Update doc comments for spawn hook.

m-ou-se commented:
@WaffleLapkin Thanks for the review! Addressed all your comments.

WaffleLapkin reviewed (approved):
r=me with green ci

m-ou-se commented:
@bors r=WaffleLapkin

m-ou-se commented:
@bors r=WaffleLapkin

m-ou-se commented:
@bors r=WaffleLapkin

bors commented:
:pushpin: Commit 691796be03d77951982a86f4913994791ed4fb61 has been approved by `WaffleLapkin`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=WaffleLapkin 691796be03d77951982a86f4913994791ed4fb61 -->
<!-- homu: {"type":"Approved","sha":"691796be03d77951982a86f4913994791ed4fb61","approver":"WaffleLapkin","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
